source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.cc

```diff
@@ -0,0 +1,264 @@
+#include "extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.h"
+
+#include <algorithm>
+#include <array>
+#include <cstdint>
+#include <deque>
+#include <functional>
+#include <string>
+#include <vector>
+
+#include "envoy/common/pure.h"
+#include "envoy/extensions/transport_sockets/tls/v3/tls_spiffe_validator_config.pb.h"
+#include "envoy/network/transport_socket.h"
+#include "envoy/registry/registry.h"
+#include "envoy/ssl/context.h"
+#include "envoy/ssl/context_config.h"
+#include "envoy/ssl/private_key/private_key.h"
+#include "envoy/ssl/ssl_socket_extended_info.h"
+
+#include "common/common/matchers.h"
+#include "common/common/regex.h"
+#include "common/config/datasource.h"
+#include "common/config/utility.h"
+#include "common/protobuf/message_validator_impl.h"
+#include "common/stats/symbol_table_impl.h"
+
+#include "extensions/transport_sockets/tls/cert_validator/factory.h"
+#include "extensions/transport_sockets/tls/cert_validator/well_known_names.h"
+#include "extensions/transport_sockets/tls/stats.h"
+#include "extensions/transport_sockets/tls/utility.h"
+
+#include "openssl/ssl.h"
+#include "openssl/x509v3.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace TransportSockets {
+namespace Tls {
+
+using SPIFFEConfig = envoy::extensions::transport_sockets::tls::v3::SPIFFECertValidatorConfig;
+
+SPIFFEValidator::SPIFFEValidator(const Envoy::Ssl::CertificateValidationContextConfig* config,
+                                 SslStats& stats, TimeSource& time_source)
+    : stats_(stats), time_source_(time_source) {
+  ASSERT(config != nullptr);
+
+  SPIFFEConfig message;
+  Config::Utility::translateOpaqueConfig(config->customValidatorConfig().value().typed_config(),
+                                         ProtobufWkt::Struct(),
+                                         ProtobufMessage::getStrictValidationVisitor(), message);
+
+  auto size = message.trust_domains().size();
+  trust_bundle_stores_.reserve(size);
+  for (auto& domain : message.trust_domains()) {
+    if (trust_bundle_stores_.find(domain.name()) != trust_bundle_stores_.end()) {
+      throw EnvoyException(absl::StrCat(
+          "Multiple trust bundles are given for one trust domain for ", domain.name()));
+    }
+
+    auto cert = Config::DataSource::read(domain.trust_bundle(), true, config->api());
+    bssl::UniquePtr<BIO> bio(BIO_new_mem_buf(const_cast<char*>(cert.data()), cert.size()));
+    RELEASE_ASSERT(bio != nullptr, "");
+    bssl::UniquePtr<STACK_OF(X509_INFO)> list(
+        PEM_X509_INFO_read_bio(bio.get(), nullptr, nullptr, nullptr));
+    if (list == nullptr || sk_X509_INFO_num(list.get()) == 0) {
+      throw EnvoyException(
+          absl::StrCat("Failed to load trusted CA certificate for ", domain.name()));
```

Comment:
Coverage report indicates there is no coverage for 4 throws in this file, plus several other error handling branches.

https://storage.googleapis.com/envoy-pr/684420f/coverage/source/extensions/transport_sockets/tls/cert_validator/default_validator.cc.gcov.html

Reply:
That is for default_validator, not for SPIFFE one. I see these are uncovered quite a while.